jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ThreadUtils.java

```
@@ -0,0 +1,39 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.thread;
+
+/**
+ * ThreadUtils provides misc. utilities for threads.
+ */
+public class ThreadUtils
+{
+    private static ThreadGroup innocuousThreadGroup;
+
+    public static boolean isInnocuous(Thread thread)
+    {
+        if (Runtime.version().feature() < 25)
+            return false;
+
+        if (innocuousThreadGroup != null)
+            return thread.getThreadGroup() == innocuousThreadGroup;
+
+        ThreadGroup threadGroup = thread.getThreadGroup();
+        if (threadGroup != null && "InnocuousThreadGroup".equals(threadGroup.getName()))
```

I am -1 on this check against private JDK classes that may be renamed, or changed.

The cases where we fail to set the TCCL are very restricted, and I would rather try/catch in those restricted cases than having to maintain this code over JDK versions.

`InnocuousThread` was introduced in JDK 9 12 years ago (https://github.com/openjdk/jdk/commit/115e92c184f1e75215cadc4b5a1100578d72b055), so the initial check about Java 25 is already possibly wrong.

---

I personally am -0 on this. I don't mind adding implementation-specific checks as long as they are totally safe, which is the case here. But I do hear the argument that this is an extra annoyance to maintain just for a special, potentially rare use-case.

---

> I personally am -0 on this. I don't mind adding implementation-specific checks as long as they are totally safe, which is the case here. 

How are they safe? 
A rename of the JDK thread group would pass under the radar, and we will be back to this issue not being resolved.
And then we would have to sniff the JDK version, or have a list of names to compare against, and hope that when a name matches, it won't be for a thread for which it would be possible to set the TCCL.

For example, what happens with virtual threads?
My suspect is that JDK's AIO has not been updated to virtual threads internally, because it probably relies on user code to pass in a virtual thread `Executor` via `AsynchronousFileChannel.open(... Executor executor, ...)` (https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/nio/channels/AsynchronousFileChannel.html#open(java.nio.file.Path,java.util.Set,java.util.concurrent.ExecutorService,java.nio.file.attribute.FileAttribute...)), at which point all the bets on the thread group name are off.

The `try/catch` will work in all cases.

---

I too an -0 on name checks.

But I'm -1 on ignoring any thrown exception on the hot path.   IF we are going to work when called by an Innocuous Thread, then we should work efficiently.     I'd much rather fail with a reason than work slowly.

If we need this method (and I currently don't think we do), we could perhaps avoid the name comparison if we could reliably create a single Innocuous Thread statically and take a copy of its ThreadGroup.

We should try to do that as we could then put some asserts into enter like:

```java
   assert _classloader == null || !ThreadUtils.isInnocuous(Thread.currentThread());
```

I  know people don't always run with asserts on, but it would give us an easier way to declare in the code where innocuous threads are not expected to work

---

Ah! looking at this more, I do not think we can implement isInnocuous, as the JVM treats innocuous threads more of a concept rather than a specific type.   There are at least two separate implementations:
 + jdk.internal.misc.InnocuousThread
 + java.util.concurrent.ForkJoinWorkerThread.InnocuousForkJoinWorkerThread

Thus there is nothing to stop future implementations of the JVM to add more Innocuous Threads implementations and/or change the names.   There is no single ThreadGroup we can check and no interface.

I think they should never have let these threads into the wild!  I get it that you might want a very restricted thread to be used in a callback in some circumstances, but what they are going to cause is that every callback that calls a 3rd party library now will have to do an extra dispatch.    That would be fine if innocuous threads were really cheap, but they are not.

I think we just need to recommend that a thread pool always be used with APIs that might otherwise use innocuous threads.

---

@lorban @sbordet see above

---

@gregw I agree about the fact that we cannot implement `isInnocuous()`.

Having said that, we have 2 choices:

1. Try to set the TCCL, catch, and hope that without the TCCL whatever code needed it works.
2. Try to fail hard.

Solution 2 seems to be quite complicated due to the fact that enterScope/exitScope are implemented for generic lambdas such as predicates, runnables, consumers, etc.
If we cannot enterScope, throwing would likely result in some unexpected behavior and the lambda would not be invoked.
For example, the predicate case is called on an idle timeout event, and throwing because we cannot enterScope would basically result in the idle timeout being ignored.

I'm for solution 2. but it requires a large review.

---

I'm for option 3. Don't set a context classloader if the context doesn't have a classloader

---

Or option 4.  Have a special NO_CLASS_LOADER constant ClassLoader that if set on a context stops it setting/resetting the TCCL - otherwise we keep our behaviour the same.
Those wanting to call us with InnocuousThreads will have to set NO_CLASS_LOADER - a PITA but perhaps better than a behaviour change.